src/wp-includes/class-wp-scripts.php

```diff
@@ -512,31 +476,125 @@ public function add_inline_script( $handle, $data, $position = 'after', $standal
-			if ( 'after-non-standalone' === $position ) {
-				$script_output = "<script%1\$s id='%2\$s-js-after' data-wp-executes-after='%2\$s'>\n%4\$s\n</script>\n";
-			} else {
-				$script_output = "<script%1\$s id='%2\$s-js-%3\$s'>\n%4\$s\n</script>\n";
-			}
+	/**
+	 * Gets tags for inline scripts registered for a specific handle.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $handle   Name of the script to add the inline script to.
+	 *                         Must be lowercase.
+	 * @param string $position Optional. Whether to add the inline script
+	 *                         before the handle or after. Default 'after'.
+	 * @return string Inline script, which may be empty string.
+	 */
+	public function get_inline_script_tag( $handle, $position ) {
+		$js = $this->get_inline_script_data( $handle, $position );
+		if ( empty( $js ) ) {
+			return '';
+		}
+
+		// TODO: Handle case where a dep has a src which is false.
+		$id   = "{$handle}-js-{$position}";
+		$deps = $this->registered[ $handle ]->deps;
 
-			printf(
-				$script_output,
-				$this->type_attr,
-				esc_attr( $handle ),
-				esc_attr( $position ),
-				$output
+		if ( $this->should_delay_inline_script( $handle, $position ) ) {
+			$attributes = array(
+				'id'   => $id,
+				'type' => 'text/template', // TODO: Consider text/plain instead.
 			);
+			if ( $deps ) {
+				$attributes['data-wp-deps'] = implode( ',', $deps );
+			}
+			return wp_get_inline_script_tag( $js, $attributes );
+		} else {
+			return wp_get_inline_script_tag( $js, compact( 'id' ) );
 		}
+	}
 
-		return $output;
+	/**
+	 * Determines whether an inline script should be delayed.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $handle   Name of the script to add the inline script to.
+	 *                         Must be lowercase.
+	 * @param string $position Optional. Whether to add the inline script
+	 *                         before the handle or after. Default 'after'.
+	 *
+	 * @return bool Whether to delay.
+	 */
+	private function should_delay_inline_script( $handle, $position ) {
+		// Never delay the inline script if the script is blocking.
+		if ( ! $this->is_delayed_strategy( $this->get_eligible_loading_strategy( $handle ) ) ) {
+			return false;
+		}
+
+		// After inline scripts must always be delayed for non-blocking scripts.
+		if ( 'after' === $position ) {
+			return true;
+		}
+
+		// From now on, we're only considering before inline scripts.
+		$deps = $this->registered[ $handle ]->deps;
+
+		/*
+		 * If there are no dependencies, the before script must not delay since there will not be a load event on a
+		 * preceding script for which an event handler can run the before inline script.
+		 */
+		if ( empty( $deps ) ) {
+			return false;
+		}
+
+		$has_blocking_dependency = false;
+		foreach ( $this->registered[ $handle ]->deps as $dep ) {
+			if ( ! $this->is_delayed_strategy( $this->get_eligible_loading_strategy( $dep ) ) ) {
+				$has_blocking_dependency = true;
+				break;
+			}
+		}
```

Comment:
Hummm. This then breaks the [blocking-with-async-dependency](https://github.com/westonruter/wp-script-loading-strategy-tests/blob/main/cases/blocking-with-async-dependency.php) test, causing a regression back to what this test case originally identified as an error:

```
 * Note: This was sometimes result in a snapshot as follows:
 *
 *   blocking-not-async-without-dependency: before inline
 *   blocking-not-async-without-dependency: script
 *   blocking-not-async-without-dependency: after inline
 *   async-with-blocking-dependency: script
 *   async-with-blocking-dependency: after inline
 *   async-with-blocking-dependency: before inline
 *
 * The before script for async-with-blocking-dependency is getting executed _after_ the after script. This appears to
 * have been fixed with the introduction of \WP_Scripts::should_delay_inline_script() which prevents delaying a before
 * script when there are blocking dependencies.
```

The delayed `before` script is not running before but rather after. Looking at the test output HTML:

```html
<script id="blocking-not-async-without-dependency-js-before">
scriptEventLog.push( "blocking-not-async-without-dependency: before inline" )
</script>
<script src='http://localhost:8889/wp-content/plugins/script-loading-strategy-tests/external.js?script_event_log=blocking-not-async-without-dependency%3A+script&#038;ver=6.3-alpha-55505-src' id='blocking-not-async-without-dependency-js'></script>
<script id="blocking-not-async-without-dependency-js-after">
scriptEventLog.push( "blocking-not-async-without-dependency: after inline" )
</script>
<!-- 👆👇 Problem right here! -->
<script id="async-with-blocking-dependency-js-before" type="text/plain" data-wp-deps="blocking-not-async-without-dependency">
scriptEventLog.push( "async-with-blocking-dependency: before inline" )
</script>
<script src='http://localhost:8889/wp-content/plugins/script-loading-strategy-tests/external.js?script_event_log=async-with-blocking-dependency%3A+script&#038;ver=6.3-alpha-55505-src' id='async-with-blocking-dependency-js' async></script>
<script id="async-with-blocking-dependency-js-after" type="text/plain" data-wp-deps="blocking-not-async-without-dependency">
scriptEventLog.push( "async-with-blocking-dependency: after inline" )
</script>
```

The problem is that there is no `load` event firing _after_ the execution of the blocking script's inline `after` script. This means there's nothing to attach to for executing the inline before script at the right point. 

So the async script's inline `before` script needs to not be delayed. And actually it's perfectly fine for such an inline `before` script to be run synchronously because there's no need to delay it. Furthermore, it can't be delayed if we rely on the `load` event of the preceding delayed external script to evaluate such inline scripts.

Reply:
OK, I reverted my change in 1e3d57bc7e2a69b56c59712679325ea7093789e6, restoring the original logic you were questioning. I improved the inline comment explaining why a `before` inline script cannot be delayed when there is a blocking dependency:

https://github.com/10up/wordpress-develop/blob/1e3d57bc7e2a69b56c59712679325ea7093789e6/src/wp-includes/class-wp-scripts.php#L580-L594

Let me know if that addresses your comment:

> This logic seems backwards to me. My understanding is that we would only return true if one of the dependencies was using a delayed strategy. If all the dependencies are blocking, then we would not need to delay the inline before script, unless I'm misunderstanding.

In other words, there is no need to ever delay an inline `before` script for a delayed dependency if it doesn't also have delayed dependencies. If it has a blocking dependency, the expected execution order will always be preserved by running the inline `before` script immediately. And with the refactor to use the `load` event on the dependency to fire the `before` inline scripts of the dependents, it turns out to be required to never delay the first delayed dependent's `before` inline script, since we need to make sure that any inline `after` scripts of the dependency have executed first.

Reply:
The explanation you have here aligns with my expectations, but I'm still struggling over the current logic as it's implemented. The way I'm reading the current logic, it assumes any inline script added before a delayed script handle will be delayed unless it encounters a blocking dependency. Instead, I would expect it to assume the inline script will not be delayed unless one of the dependencies is delayed. I.e.:

```php
foreach ( $deps as $dep ) {
	if ( $this->is_delayed_strategy( $this->get_eligible_loading_strategy( $dep ) ) ) {
		return true;
	}
}

return false;
```

Reply:
@joemcgill I had to add some more "E2E" test cases to figure this out. And the difference between my code and your suggestion seems to come down to when a delayed dependency has one blocking dependency and one delayed dependency.

Version | [defer-dependent-with-blocking-and-defer-dependencies](https://github.com/westonruter/wp-script-loading-strategy-tests/blob/main/cases/defer-dependent-with-blocking-and-defer-dependencies.php) | [defer-dependent-with-defer-and-blocking-dependencies](https://github.com/westonruter/wp-script-loading-strategy-tests/blob/main/cases/defer-dependent-with-defer-and-blocking-dependencies.php)
---|---|---
Mine | ![image](https://github.com/10up/wordpress-develop/assets/134745/ebb43ce6-13c5-4c75-8cd8-321e23730697) | ![image](https://github.com/10up/wordpress-develop/assets/134745/e3d4b4df-7909-421b-b660-a0ddeccbe495)
Yours | ![image](https://github.com/10up/wordpress-develop/assets/134745/aa8d6157-dbb1-48ef-b2d6-347909ea7d68) | ![image](https://github.com/10up/wordpress-develop/assets/134745/32c656fc-3d57-4bd7-a536-102a8b9740a4)

So it seems yours does work better. In my version, the dependency script and inline after script are erroneously running before the dependent's before script. I'm not exactly clear on why that is yet.

Reply:
In the case of defer-dependent-with-defer-and-blocking-dependencies, your change has the following effect on the markup:

```diff
 <script id="blocking-dependency-with-defer-preceding-dependency-js-after" type="text/javascript"  >
 scriptEventLog.push( "blocking-dependency-with-defer-preceding-dependency: after inline" )
 </script>
-<script id="defer-dependent-of-defer-and-blocking-dependencies-js-before" type="text/javascript"  >
+<script data-wp-deps="defer-dependency-with-blocking-following-dependency,blocking-dependency-with-defer-preceding-dependency" id="defer-dependent-of-defer-and-blocking-dependencies-js-before" type="text/plain"   >
 scriptEventLog.push( "defer-dependent-of-defer-and-blocking-dependencies: before inline" )
 </script>
 <script defer id="defer-dependent-of-defer-and-blocking-dependencies-js" src="https://example.com/external.js?script_event_log=defer-dependent-of-defer-and-blocking-dependencies%3A+script&amp;ver=6.3-alpha-55505-src" type="text/javascript"    ></script>
```

It's not clear to me why this works because it would seem that like this would not give an opportunity for the inline `before` script be evaluated before the corresponding dependency is loaded. I'll need to sleep on it.

Reply:
OK, thinking about it more, the reason your suggestion works is that when there is a delayed dependency, there _will_ be an opportunity at a `load` event to fire the `before` script at the right time. In contrast, my version forces the `before` script of a dependent to run too early, before any delayed dependency script has run.